### DISCUSSION TITLE
Graphcache not returning partial results

### DISCUSSION THREAD

**Discussion by thisiszinger** [urql-graphql/urql#2708] [Q&A] [answered]
Hi team! First off, thanks for this library and the care that you all put into it ❤️ I've enjoyed using `urql` (coming from an Apollo Client background) and learning more about how it works in the process. I'm currently migrating to Graphcache and want to take advantage of the partial results feature. In my case, I have a list view and a detail view with the following queries:

```gql
query AllReleases {
  releasesConnection {
    __typename
    nodes {
      __typename
      id
      title
    }
  }
}

query ReleaseById($id: String!) {
  releaseById(id: $id) {
    __typename
    ... on Release {
      __typename
      id
      title
      topBid {
        __typename
        id
        amount
      }
    }
  }
}
```

When I click into the detail view after loading the list view, none of the data is displayed until the request is completed. I would expect that `id` and `title` resolve immediately from the cache and `topBid` is added once the request completes. It is working as expected if I remove `topBid` from the query, but my understanding based on the docs and discussions here is that cached data should return immediately so long as there aren't any non-null fields missing from the cache. Thinking that it might be related to non-null objects, I've also tried adjusting the detail view query to include a non-null scalar instead but am still experiencing the same behavior.

```gql
query ReleaseById($id: String!) {
  releaseById(id: $id) {
    __typename
    ... on Release {
      __typename
      id
      title
      publishedAt
    }
  }
}
```

I've generated my schema (`graphql.schema.json`) using [@graphql-codegen/urql-introspection](https://npmjs.com/package/@graphql-codegen/urql-introspection) and have passed it into `cacheExchange`. I should note that I am getting a TypeScript error here (`Types of property 'kind' are incompatible. Type 'string' is not assignable to type '"INTERFACE"'.`) but wasn't sure if it would affect the functionality. `releaseById` returns a union (hence the `... on Release` fragment) and I'm unclear if that's the issue. I'm also using [@graphql-codegen/typescript-urql-graphcache](https://npmjs.com/package/@graphql-codegen/typescript-urql-graphcache) to generate a `GraphCacheConfig`, but I don't believe that should make a difference (I still see the same error if I remove that). My `cacheExchange` looks like this:

```
export const cacheExchange = graphcacheExchange<GraphCacheConfig>({
  // @ts-expect-error
  schema,
  resolvers: {
    Query: {
      releaseById: (_, args) => ({ __typename: "Release", id: args.id }),
      userById: (_, args) => ({ __typename: "User", id: args.id }),
    },
  },
});
```

My main questions are:

1. Why isn't the cached, non-null data resolving immediately?
2. How can I fix the type error related to the schema?
3. Can I remove some of those `__typename`s from my queries (I started adding them everywhere to see if that would fix the issue, still a bit confused about when they're necessary vs. not)?
4. Is it recommended to use [@graphql-codegen/urql-introspection](https://npmjs.com/package/@graphql-codegen/urql-introspection) over [@graphql-codegen/introspection](https://npmjs.com/package/@graphql-codegen/introspection) assuming that this is the only place that the schema is being used?
5. Should the schema file be committed to git? I'm unclear how it would be generated in a production environment with introspection disabled (is it even needed in production?).

Thanks again for your time!

**Comment by JoviDeCroock** [comment, marked as answer]
Thank you so much for the kind words!

1. non-null data depends on your schema, so things annotated with `!` have to be fetched and are considered to be needed for your UI as the typings of the schema tell us so, there is a proposal in GraphQL for client controlled nullability but not there yet. [link](https://github.com/graphql/graphql-wg/issues/694)
2. I personally just do `as any` because it's a side-effect of the minimized json structure
3. urql adds this itself so you can omit them if you don't need them in the UI as data
4. the `urql-introspection` is more bundle-size efficient as we strip away everything we don't need
5. I commit/add it to my publishes

**Reply by thisiszinger** [reply]
1. That's my confusion: both `topBid` and `publishedAt` are nullable in the schema. I read through a bunch of other discussions surrounding this and most of them were resolved because there was a misunderstanding and the fields were actually non-null, but in my case, those are both definitely nullable yet they are preventing the partial data from being returned.
2. Got it, thanks!
3. When I omit `__typename` I'm seeing the following error in the console

```
Couldn't find __typename when writing.
If you're writing to the cache manually have to pass a `__typename` property on each entity in your data.
(Caused At: "RefreshToken" mutation)
```

For reference, the `RefreshToken` mutation looks like this:

```gql
mutation RefreshToken {
  refreshToken {
    success
    accessToken
    problems {
      message
    }
  }
}
```

And I've also updated my Graphcache config as follows since the types returned by that mutation don't have IDs and aren't important to store in the cache:

```ts
export const cacheExchange = graphcacheExchange<GraphCacheConfig>({
  schema: schema as any, // type conflict due to minified schema
  keys: {
    InvalidInputProblem: () => null,
    NotFoundProblem: () => null,
    RefreshTokenPayload: () => null,
    UnauthorizedProblem: () => null,
  },
  resolvers: {
    Query: {
      releaseById: (_, args) => ({ __typename: "Release", id: args.id }),
      userById: (_, args) => ({ __typename: "User", id: args.id }),
    },
  },
});
```

4. Sounds good, will stick with that! 
5. Good to know, will try the same!

**Reply by JoviDeCroock** [reply]
1. If that's a bug and you can reproduce it definitely open to research that in an issue given a reproduction.
3. Hmm, normally we add `__typename`  everywhere weird that it doesn't here, this normally happens with the [formatDocument helper](https://github.com/FormidableLabs/urql/blob/57894b9a11aa28634ec3e0e5f31d616261e5e043/packages/core/src/utils/typenames.ts#L70) but maybe we have a missing case I will check this out.

**Comment by probablykabari** [comment]
@zingerj unrelated to your issue but just noticing the `ts-expect-error`. You can update your codgen config to export a typescript file that you can import  so you don't need that:

```
 // this will make it export IntrospectionQuery type instead of raw json
  ./my/path/urql.introspection.ts:
    config:
      module: 'es2015'
      useTypeImports: true
    plugins:
      - "urql-introspection"

// in urql client, no type errors!
import schema from '../my/path/urql.introspection'

```

You can control minification in your normal tsc/vite/whatever build process 😃

**Reply by thisiszinger** [reply]
Hmm okay, I'll look into that. Doesn't `urql-introspection` already minify the output?

**Reply by probablykabari** [reply]
There might be a setting I missed but mine does not minify until bundling for prod.

**Reply by thisiszinger** [reply]
Are you doing anything specific to minify this file? Or do you just mean it's being treated like the rest of your TS files?

**Reply by probablykabari** [reply]
Treating it like the rest of TS.
